grammars/prql-lezer/src/prql.grammar

```diff
@@ -3,114 +3,153 @@
 //   normal functions?
 // - A few small TODOs included below
 
+@precedence {
+  times @left,
+  plus @left,
+  compare @left,
+  and @left,
+  or @left
+}
+
 @top Query { Statements }
-@skip { space | Comment | wrapped_line }
 
-Statements { newline* Query_def? (Def)* Pipeline_stmt? end }
+@skip { space | Comment | Docblock | wrappedLine }
+
+Statements { newline* QueryDefinition? VariableDeclaration* PipelineStatement? end }
 
-Query_def { @specialize<ident_part, "prql"> Named_arg* newline+ }
+QueryDefinition { @specialize<identPart, "prql"> NamedArg+ newline+ }
 
-Pipeline_stmt { Pipeline ( ~ambig_newline newline+ | end )}
+PipelineStatement { Pipeline ( ~ambigNewline newline+ | end )}
 
-Pipeline { expr_call (pipe+ expr_call)* }
+Pipeline { exprCall (pipe+ exprCall)* | test }
 
-pipe { "|" | ~ambig_newline newline }
+pipe { "|" | ~ambigNewline newline }
 
-Tuple { "{" newline* tuple_item (("," newline* ) tuple_item)* ","? newline* "}" }
+TupleExpression { "{" newline* tupleItem (("," newline* ) tupleItem)* ","? newline* "}" }
 
-tuple_item { Assign_call | expr_call | Case_branch }
+tupleItem { AssignCall | exprCall | CaseBranch }
 
-expr_call { Expr | Func_call }
+exprCall { expression | CallExpression }
 // Ideally we would force a space after `Ident` to prevent an invalid s-string
-// being parsed as a Func_call, e.g. `s"{a"` -> `s` & `"{a"'. But we
+// being parsed as a CallExpression, e.g. `s"{a"` -> `s` & `"{a"'. But we
 // can't seem to force a space because it's in our skip, and I can't see
 // a way of changing the skip expression to only specialize on a single item
-Func_call { Ident ( (Named_arg | Assign | Expr))+ }
+CallExpression { Ident ( (NamedArg | Assign | expression))+ }
 
-// !term is required here, otherwise `[x+emp_no, -3]` is parsed as `x` & `+emp_no`
-term { !term Ident | number | Tuple | DateTime | Nested_pipeline | op_unary term | String | R_string | S_string | F_string | Range | Array | Case }
+NamedArg { identPart ":" expression }
+Assign { identPart "=" expression }
+AssignCall { identPart "=" exprCall }
+CaseBranch { expression "=>" expression }
+// Possibly we could only accept case branches inside the TupleExpression?
+CaseExpression { @specialize<identPart, "case"> TupleExpression }
 
-Expr { term (( Op_bin ) term )* }
+NestedPipeline { "(" newline* Pipeline ~ambigNewline newline? ")" }
 
-Named_arg { ident_part ":" Expr }
-Assign { ident_part "=" Expr }
-Assign_call { ident_part "=" expr_call }
-Case_branch { Expr "=>" Expr }
-Array { "[" newline* Expr (("," newline* ) Expr)* ","? newline* "]" }
-// Possibly we could only accept case branches inside the Tuple?
-Case { @specialize<ident_part, "case"> Tuple }
+commaSep<expr> { newline* expr ("," newline* expr newline*)* ","? }
+LogicOp<expr> { expr }
+
+test {
+ testInner
+}
 
-Nested_pipeline { "(" newline* Pipeline ~ambig_newline newline? ")" }
+testInner { binaryTest | unaryTest | expression }
```

Comment:
How come these are all called `test`?

Reply:
Those are called coincidentally called `test` just because that is what they're called in the [Python grammar](https://github.com/lezer-parser/python/blob/main/src/python.grammar). They're not called that in the JavaScript or Rust parser though, so we could rename it to something else. Lezer is written in JavaScript, and the JavaScript  grammar have the most "stars" on GitHub, so I consider it the most authoritative.

Reply:
OK, maybe add a short comment saying that's why they're called `Test`, so it's not interpreted as something fundamentally different from the other tokens. Thanks...

Reply:
Okay, I added a comment about this.